### ISSUE TITLE
Strange behaviours in dual monitor setup

### ISSUE THREAD

**Issue by unman** [QubesOS/qubes-issues#6268]
**Qubes OS version**

4.0 updated

**Affected component(s) or functionality**

Qubes gui

**Brief summary**

When working with docked x230 I see strange behaviour with windows that seems new to me.
These relate to window placement and mouse registration. (I don't use mouse a great deal, so this may have been there for some time, but I think not.)

**How Reproducible**

100%

**To Reproduce**

1. Connect laptop to docking station with HDMI to external 1020x1080 monitor.

**Expected behavior**

Examples from previous behaviour:
Windows can be moved from one screen to another, and will (e.g) maximise to current screen size.
Windows will appear appropriately to current screen.
Moving window from one screen to another will leave window in usable state.

**Actual behavior**

Current behaviour:
Windows can be moved from one screen to another, but will (e.g) maximise to the wrong screen size.
Windows will appear sized for larger screen, even when placed on smaller screen.
Moving window from larger screen to smaller will leave window in strange state: mouse movements appear correct, but window and tab interactions are "off". Sometimes tab and click interactions will not work at all.
Moving the window back to the originating window generally fixes these issues.

**Screenshots**

**Additional context**

The x230 has resolution 1366x768
The HDMI connected screen 1920x1080

These problems did/do not arise using VGA connected screen.
As I'm not a great mouse user, or really of anything with significant mouse interaction, it's *possible* that these issues have been there for some time, but my feeling is that it's a more recent development.

**Solutions you've tried**

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

Couldnt find anything that seemed appropriate.

**Comment by SaswatPadhi** [comment]
I have noticed this on my machine too. This only appears on monitors that are connected after booting.

I usually restart my running VMs (those that have GUI apps running), to make them "recognize" the newly connected monitor(s).

**Comment by unman** [comment]
On Fri, Dec 11, 2020 at 05:07:05AM -0800, Saswat Padhi wrote:
> I have noticed this on my machine too. This only appears on monitors that are connected after booting.
> 
> I usually restart my running VMs (those that have GUI apps running), to make them "recognize" the newly connected monitor(s).
> 

Yes, probably when I dock the laptop.
BUT, this is new behaviour afaik

**Comment by gnorsilva** [comment]
I also see this on my machine too, eg:

- laptop monitor connected, at 1920x1080 resolution
- start a VM
- connect external ultrawide monitor at 3440x1440, disable laptop monitor
- VM apps do not register mouse interactions beyond the original 1920x1080 window area (i.e. top-left quadrant of screen)

Restarting the VM fixes the issue. Does a VM receive mouse interaction bounds only at startup time?

**Comment by andrewdavidwong** [comment]
Is this still a problem in 4.1?

**Comment by gnorsilva** [comment]
Yes, this is still an issue in 4.1

**Comment by UndeadDevel** [comment]
Confirmed that this is still an issue on 4.2.

Easiest way to reproduce in my use case is to:

1. Have a web-connected qube running
2. Connect an external monitor over HDMI, which has greater resolution than the "main monitor"
3. In the web-connected qube browse to a website with links all over the site and maximize the window
4. Observe that links outside the rectangle defined by the "main monitor"'s resolution and anchored at 0,0 cannot be opened by mousing over and clicking them; in fact, a mouse over doesn't highlight them at all (the way links are usually highlighted when mousing over them); all the while, the bottom panel, which is also outside the rectangle, continues to work fine and other mouse actions, such as scrolling, also work even in that qube outside the rectangle.

The fix seems fairly reliable: restart the qube after the external monitor was connected.

**Comment by neoniobium** [comment]
> Confirmed that this is still an issue on 4.2.
> 
> Easiest way to reproduce in my use case is to:
> 
>     1. Have a web-connected qube running
> 
>     2. Connect an external monitor over HDMI, which has greater resolution than the "main monitor"
> 
>     3. In the web-connected qube browse to a website with links all over the site and maximize the window
> 
>     4. Observe that links outside the rectangle defined by the "main monitor"'s resolution and anchored at 0,0 cannot be opened by mousing over and clicking them; in fact, a mouse over doesn't highlight them at all (the way links are usually highlighted when mousing over them); all the while, the bottom panel, which is also outside the rectangle, continues to work fine and other mouse actions, such as scrolling, also work even in that qube outside the rectangle.
> 
> 
> The fix seems fairly reliable: restart the qube after the external monitor was connected.

I can 100% reproduce this on `4.2` using either my `X230` or `T430` with a docking station via a displayport connection. I happens on `i3wm` as well as on `xfce4`.

**Comment by marmarek** [comment]
Yes, this is kinda expected behavior (aka known issue): https://www.qubes-os.org/doc/gui-configuration/#video-ram-adjustment-for-high-resolution-displays
Adding detection for this at least to show a sensible notification about it would be nice, @andrewdavidwong do we have an issue about it already?

**Comment by UndeadDevel** [comment]
@marmarek There may also be either a documentation bug or another kind of bug (quote from your linked doc article):
```
The amount of memory allocated per qube is the maximum of:

   - gui-videoram-min
   - current display + gui-videoram-overhead

Default overhead is about 8 MiB, which is enough for a 1080p display (see above). So, the gui-videoram-overhead zeroing is not strictly necessary; it only avoids allocating memory that will not be used.
```
This suggests that by default the video RAM allocated per qube should be about 16 MB if the "main monitor" (always connected) is 1080p (as "current display" would be $((1920 * 1080 * 4 / 1024)) = 8100 and default overhead is supposed to be 8 MB, so in total 8 + 8.1 = 16.1 MB).

16.1 MB would be enough for a QHD monitor (1440p) alone (my use case), because here the required RAM would be $((2560 * 1440 * 4 / 1024)) = 14400 (14.4 MB), but I'm still seeing the bug.

I should also note that `qvm-features dom0` did not return any of the two mentioned properties (`gui-videoram-min` and `gui-videoram-overhead`).

Executing `qvm-features dom0 gui-videoram-min 14400` fixes the issue in my case (i.e. when using the external 1440p monitor as the only monitor).

**Comment by marmarek** [comment]
It may depend on the arrangement of the displays, as the whole framebuffer is always rectangle, even if you have different screen sizes. See `xrandr` (the top number) for the actual size.

**Comment by UndeadDevel** [comment]
The `xrandr` values are as expected; I meant that the documentation seems to be misleading or it doesn't work as described (or I'm misunderstanding the docs), because in my understanding the default ram size per qube with a 1080p monitor should be ~16 MB (8 for the display plus 8 overhead), which should be enough (with minimal leftover overhead) for an external 1440p display if connected and used as the sole display (it needs 14.4 MB) and yet the issue occurs.

After some more testing I noticed that increasing videoram min size only fixes the issue if the external monitor is used alone; when using it as an additional monitor (extension to the right) then no amount of videoram assignment will fix the issue, which now affects that qube across the entirety of the (additional, external) screen; only a reboot of the affected qube will fix it then.

**Comment by marmarek** [comment]
Note the value set in `gui-videoram-min` takes effect after restarting a qube only.

**Comment by UndeadDevel** [comment]
Hmm; maybe I inadvertently rebooted that qube as it was one of my cached DVMs...

In any case, as thankfully the `gui-videoram-min` value persists across dom0 reboots, I see that this issue is fixed for me now, including in extended mode, with a value of 25200 for the mentioned property, without having to restart VMs.

Thanks marmarek for the pointer to the docs and explanations!

**Comment by pefu** [comment]
I found this issue today. A long time QubesOS user asked a question in our german language QubesOS telegram user chat group. He observed in QubesOS 4.2 this behaviour when moving windows from his internal laptop screen to later plugged in external monitor and asked whether in 4.2 this is a new problem. I tried to reproduce with my current QubesOS 4.1 and indeed: I had to restart any App qube which was running before the external monitor was plugged in. Otherwise mouse input was impossible in windows on the "new" screen "real estate". I just did not notice this bug because I normally always start my laptop with the external monitor already plugged in.

**Comment by observinglynx** [comment]
I'm running into the issue pefu mentioned here, perhaps this is also the same issue that unman reported.
I've reproduced it on 4.0-4.2, it occurs on qubes that have been started before new display(s) are connected.

For example, with 2x `1920x1080` displays connected before the qube is started and then starting the qube and executing `xrandr` in it, you get something like this in response:
```
DUMMY0
   QB3840x1080...
```
But if you start the qube with only one display connected and then connect a second display after the qube has started, you get two dummy outputs:
```
DUMMY0
   QB1920x1080...
DUMMY1
   QB1920x1080...
```
The problem here seems to be that the DUMMY1 output does not get the correct position offset:
```
Screen 0: minimum 64 x 64, current 1920 x 1080, maximum 32767 x 32767
DUMMY0 connected primary 1920x1080+0+0 325mm x 182mm
   QB1920x1080   46.10*+
DUMMY1 connected 1920x1080+0+0 0mm x 0mm
   QB1920x1080   46.10*+  46.10*
```
Which would mean that the outputs are stacked on top of each-other and while you can move the window _visually_ to the other display (X server on dom0 knows what the real positions and areas are), the interactable area for the qube remains on the first display.

I have worked around these issues in two ways if I would like to avoid restarting a qube(also a valid workaround):
- Generate a new mode for dummy0 with `gtf` (use the web for further instructions)
```bash
user:~$ gtf $((1920*2)) 1080 46.10
```
- Set the position offset on dummy1
```bash
user:~$ xrandr --output DUMMY1 --pos 1920x0
```

Bottom line, additional displays do not appear to receive the offsets in the qube(s) after you configure them on dom0.

**Comment by marmarek** [comment]
> Yes, this is kinda expected behavior (aka known issue): https://www.qubes-os.org/doc/gui-configuration/#video-ram-adjustment-for-high-resolution-displays

@observinglynx have you done this step?

**Comment by observinglynx** [comment]
@marmarek I tested it and it does resolve the issue, with the position offsets now being correctly set, thanks! :metal: 
Initially when I read the doc, I had misunderstood what it was for and thus overlooked it. :)

**Comment by ben-grande** [comment]
Closing as it seems to have been resolved, per the last user reports. If anybody is still facing it, someone will reopen this issue.
